driver/src/main/java/org/neo4j/driver/internal/async/pool/ConnectionPoolImpl.java

```diff
@@ -165,7 +166,10 @@ public int idleConnections( BoltServerAddress address )
             }
             finally
             {
-                eventLoopGroup().shutdownGracefully();
+                // This is an attempt to speed up the shut down procedure of the driver
+                // Feel free return this back to shutdownGracefully() method with default values
+                // if this proves troublesome!!!
+                eventLoopGroup().shutdownGracefully(200, 15_000, TimeUnit.MILLISECONDS);
```

Comment:
What does this do? Add a timeout? What happens if the timeout expires... what is the user supposed to do?

I am very cautious about optimising for a very niche use case. We need to optimise for the general case, wherein the driver shuts down at application shutdown. Here, everything _must_ close; there is no other code path.

Reply:
A short explanation how the shutdownGracefully works. It has 2 parameters ‘quiet period’ and  ‘timeout’. The algorithm that it uses is the following:

shutdownStart = now()
while (true) {
	Thread.sleep(quietPeriod)
	if (nothingWasSubmittedToTheEventLoopWhileSleeping()) {
           break;
	}

	if (now() - shutdownStart > timeout) {
          // timeout
	  break;
        }
}

The main purpose of this shutdown algorithm is to compete work that is in progress.
The default value for ‘quiet period’  are 2 seconds and 15 seconds for ‘timeout’

What do we need a quiet period for anyway? The current shutdown logic is like this:

channels.each (ch -> closeAsync())
shutdownEventLoopWithQuietPeriod()

The graceful shutdown with the quiet period is needed because closing a channel async generates events submitted to the event loop.

What we would need is something like this:
channels.each (ch -> closeAsync())
waitForAllChannlesToBeClosed()
shutdownEventLoopWithoutQuietPeriod()

As simple as the pseudo code above looks, the problem is that achieving waitForAllChannlesToBeClosed() with current Netty API is not that simple (at least was not the last time I checked)

The problem is that the shutdown currently takes 4s (the quiet period is usually reset once by the tasks generated by closing the channels) which is enough to make for instance lambda users or users with a lot of integration tests sad.

The reasoning behind my PR was that the defaults in Netty are set so it would work with all possible use cases the toolkit is used for, but 200ms quiet period is more that enough for us.